gadget/update.go

```diff
@@ -686,7 +687,16 @@ func DiskTraitsFromDeviceAndValidate(expLayout *LaidOutVolume, dev string, opts
 			continue
 		}
 
-		part, ok := diskPartitionsByOffset[uint64(structure.StartOffset)]
+		// Use the matched disk partition to get the disks.Partition data
+		// (StartOffset from LaidOutStructure is not necessarily the same
```

Comment:
it feels like we should update the comment on StartOffset in LaidOutStructure to explain this better, do we have other places using StartOffset that are potentially problematic?   also it seems that maybe we need variant of EnsureVolumeCompatibility that operates directly on LaidOutVolume and updates it? the new code here seems matching/symmetric with searchVolumeWithTraitsAndMatchParts, is that correct?

Reply:
I have added a comment about this in `LaidOutStructure.OnDiskStructure`. Usage of the OnDiskStructure from LaidOutStructure is problematic and it has been taken into account in most places already, this one was an oversight. In the end what I have wanted to do for a while is to create LaidOutStructure only when we already know the matching OnDiskStructure, but that refactoring has not been finalized yet. I prefer to keep EnsureVolumeCompatibility as it is for the moment for this reason.

Good point about searchVolumeWithTraitsAndMatchParts, in the end I have created a helper function to set OnDiskStructure which can be used here as well.